File: Arrays/Longest-Subarray-with-Sum-K/Python/Longest-Subarray-with-Sum-K-Optimal.py
"""
Problem: Longest Subarray with Sum K
Approach: Optimal - Sliding Window

This approach works when all array elements are non-negative.

Expand the window using right.
If the sum becomes greater than k, shrink from the left.

Time Complexity: O(n)
Space Complexity: O(1)
"""

class Solution:
    def longestSubarray(self, nums: list[int], k: int) -> int:
        left = 0
        curr_sum = 0
        length = 0

        for right in range(len(nums)):
            curr_sum += nums[right]

            while curr_sum > k and left <= right:
                curr_sum -= nums[left]
                left += 1

            if curr_sum == k:
                length = max(length, right - left + 1)

        return length

File: Arrays/Longest-Subarray-with-Sum-K/JavaScript/Longest-Subarray-with-Sum-K-Optimal.js
/**
 * Problem: Longest Subarray with Sum K
 * Approach: Optimal - Sliding Window
 *
 * Works when all elements are non-negative.
 *
 * Time Complexity: O(n)
 * Space Complexity: O(1)
 */

/**
 * @param {number[]} nums
 * @param {number} k
 * @return {number}
 */
var longestSubarray = function(nums, k) {
    let left = 0;
    let currSum = 0;
    let length = 0;

    for (let right = 0; right < nums.length; right++) {
        currSum += nums[right];

        while (currSum > k && left <= right) {
            currSum -= nums[left];
            left++;
        }

        if (currSum === k) {
            length = Math.max(
                length,
                right - left + 1
            );
        }
    }

    return length;
};